tests/test_loaders.py

```diff
@@ -200,6 +200,27 @@ def test_csv_pandas_column_unsupported_data_types(self, tmp_path, noaa_jfk_schem
         for t in (err_column.dtype, err_column.check):
             assert re.search(rf"{t}(\d*|ing)\b", str(e.value))  # "ing" is for "str'ing'"
 
+    def test_csv_pandas_delimiter(self, tmp_path, noaa_jfk_schema):
+        "Test setting delimiters."
+
+        def load_data():
+            return Dataset(noaa_jfk_schema, tmp_path,
+                           mode=Dataset.InitializationMode.DOWNLOAD_AND_LOAD).data['jfk_weather_cleaned']
+
+        data = load_data()
+        assert len(data.columns) == 16  # repeat partly test_csv_pandas_loader here for more clarity
+
+        # Remove the delimiter option
+        del noaa_jfk_schema['subdatasets']['jfk_weather_cleaned']['format']['options']['delimiter']
+        data = load_data()
+        assert len(data.columns) == 16  # Number of columns remain the same
+
+        # Change delimiter to tab
+        noaa_jfk_schema['subdatasets']['jfk_weather_cleaned']['format']['options']['delimiter'] = '\t'
+        del noaa_jfk_schema['subdatasets']['jfk_weather_cleaned']['format']['options']['columns']
+        data = load_data()
+        assert len(data.columns) == 1   # No tab in the file, number of columns should be 1
```

Comment:
Are there any other delimiters we should test (e.g. ` `, `;`, `|`)? Delimited files are probably the most common dataset type we'll get I'm guessing, so it might be wise to test this a little more rigorously unless you had something else in mind.

Reply:
It's a good call. I expanded the tests, but they are failing for space. I think it is related to some strange pandas behavior:

```python
In [2]: pd.read_csv('noaa-weather-data-jfk-airport/jfk_weather_cleaned.csv', delimiter=' ', encoding='utf-8', parse_dates=False)
Out[2]:
           DATE,HOURLYVISIBILITY,HOURLYDRYBULBTEMPF,HOURLYWETBULBTEMPF,HOURLYDewPointTempF,HOURLYRelativeHumidity,HOURLYWindSpeed,HOURLYStationPressure,HOURLYSeaLevelPressure,HOURLYPrecip,HOURLYAltimeterSetting,HOURLYWindDirectionSin,HOURLYWindDirectionCos,HOURLYPressureTendencyIncr,HOURLYPressureTendencyDecr,HOURLYPressureTendencyCons
2010-01-01  01:00:00,6,33,32,31,92,0,29.97,29.99,0.01,29.9...                                                                                                            
2010-01-01  02:00:00,6,33,33,32,96,0,29.97,29.99,0.02,29.9...                                                                                                            
2010-01-01  03:00:00,5,33,33,32,96,0,29.97,29.99,0,29.99,0...                                                                                                            
2010-01-01  04:00:00,5,33,33,32,96,0,29.95,29.97,0,29.97,0...                                                                                                            
2010-01-01  05:00:00,5,33,32,31,92,0,29.93,29.96,0,29.95,0...                                                                                                            
...                                                       ...                                                                                                            
2018-07-27  19:00:00,10,76,73,72,88,3,30,30.02,0,30.02,-0....                                                                                                            
2018-07-27  20:00:00,4,69,69,69,100,13,29.99,30.01,1.16,30...                                                                                                            
2018-07-27  21:00:00,10,71,70,70,96,0,30.02,30.04,0.01,30....                                                                                                            
2018-07-27  22:00:00,10,72,71,70,94,5,30,30.02,0,30.02,0.7...                                                                                                            
2018-07-27  23:00:00,10,72,71,71,97,0,30.01,30.03,0,30.03,...                                                                                                            

[75119 rows x 1 columns]

In [3]: pd.read_csv('noaa-weather-data-jfk-airport/jfk_weather_cleaned.csv', delimiter=' ', encoding='utf-8', parse_dates=[])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-618cd7665132> in <module>
----> 1 pd.read_csv('noaa-weather-data-jfk-airport/jfk_weather_cleaned.csv', delimiter=' ', encoding='utf-8', parse_dates=[])

~/wsrc/dax-api/.tox/dev/lib64/python3.8/site-packages/pandas/io/parsers.py in read_csv(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)
    686     )
    687
--> 688     return _read(filepath_or_buffer, kwds)
    689
    690

~/wsrc/dax-api/.tox/dev/lib64/python3.8/site-packages/pandas/io/parsers.py in _read(filepath_or_buffer, kwds)
    458
    459     try:
--> 460         data = parser.read(nrows)
    461     finally:
    462         parser.close()

~/wsrc/dax-api/.tox/dev/lib64/python3.8/site-packages/pandas/io/parsers.py in read(self, nrows)
   1196     def read(self, nrows=None):
   1197         nrows = _validate_integer("nrows", nrows)
-> 1198         ret = self._engine.read(nrows)
   1199
   1200         # May alter columns / col_dict

~/wsrc/dax-api/.tox/dev/lib64/python3.8/site-packages/pandas/io/parsers.py in read(self, nrows)
   2198                     values = data.pop(self.index_col[i])
   2199
-> 2200                 values = self._maybe_parse_dates(values, i, try_parse_dates=True)
   2201                 arrays.append(values)
   2202

~/wsrc/dax-api/.tox/dev/lib64/python3.8/site-packages/pandas/io/parsers.py in _maybe_parse_dates(self, values, index, try_parse_dates)
   2259
   2260     def _maybe_parse_dates(self, values, index, try_parse_dates=True):
-> 2261         if try_parse_dates and self._should_parse_dates(index):
   2262             values = self._date_conv(values)
   2263         return values

~/wsrc/dax-api/.tox/dev/lib64/python3.8/site-packages/pandas/io/parsers.py in _should_parse_dates(self, i)
   1567             else:
   1568                 name = None
-> 1569             j = self.index_col[i]
   1570
   1571             if is_scalar(self.parse_dates):

TypeError: 'NoneType' object is not subscriptable

In [4]: pd.read_csv('noaa-weather-data-jfk-airport/jfk_weather_cleaned.csv', delimiter='|', encoding='utf-8', parse_dates=[])
Out[4]:
      DATE,HOURLYVISIBILITY,HOURLYDRYBULBTEMPF,HOURLYWETBULBTEMPF,HOURLYDewPointTempF,HOURLYRelativeHumidity,HOURLYWindSpeed,HOURLYStationPressure,HOURLYSeaLevelPressure,HOURLYPrecip,HOURLYAltimeterSetting,HOURLYWindDirectionSin,HOURLYWindDirectionCos,HOURLYPressureTendencyIncr,HOURLYPressureTendencyDecr,HOURLYPressureTendencyCons
0      2010-01-01 01:00:00,6,33,32,31,92,0,29.97,29.9...                                                                                                                 
1      2010-01-01 02:00:00,6,33,33,32,96,0,29.97,29.9...                                                                                                                 
2      2010-01-01 03:00:00,5,33,33,32,96,0,29.97,29.9...                                                                                                                 
3      2010-01-01 04:00:00,5,33,33,32,96,0,29.95,29.9...                                                                                                                 
4      2010-01-01 05:00:00,5,33,32,31,92,0,29.93,29.9...                                                                                                                 
...                                                  ...                                                                                                                 
75114  2018-07-27 19:00:00,10,76,73,72,88,3,30,30.02,...                                                                                                                 
75115  2018-07-27 20:00:00,4,69,69,69,100,13,29.99,30...                                                                                                                 
75116  2018-07-27 21:00:00,10,71,70,70,96,0,30.02,30....                                                                                                                 
75117  2018-07-27 22:00:00,10,72,71,70,94,5,30,30.02,...                                                                                                                 
75118  2018-07-27 23:00:00,10,72,71,71,97,0,30.01,30....                                                                                                                 

[75119 rows x 1 columns]
```

Reply:
I've reported this bug: https://github.com/pandas-dev/pandas/issues/38489

Reply:
hm strange bug